### PR TITLE
Add option to exclude files from testing

### PR DESCRIPTION
**Pull request by midokura-silvia**
As described in #1, there is currently no way to exclude files from testing.
This pull request implements a simple way of excluding files from the command line via substring matching.

### CODE REVIEW

File: treon/treon.py [hunk 1]
```diff
@@ -127,4 +128,6 @@ def get_notebooks_to_test(args):
     if not result:
         sys.exit('No notebooks to test in {path}'.format(path=path))
 
+    for exclude_str in args['--exclude']:
+        result = [file for file in result if exclude_str not in file]
```
Comment (amit1rrr) [inline treon/treon.py]: If we accept full file & directory paths then we can have a stricter condition i.e. exclude from test if filename starts with exclude_str. 

Stricter condition is preferred to make sure only the intended files & directories are excluded.   Current pattern matching ("in" condition) can cause some unintended files to be skipped. E.g. I exclude "predictions.ipynb" from my test suite without knowing there is "final_important_predictions.ipynb" somewhere deeply nested.

File: treon/treon.py [hunk 2]
```diff
@@ -127,4 +128,6 @@ def get_notebooks_to_test(args):
     if not result:
         sys.exit('No notebooks to test in {path}'.format(path=path))
 
+    for exclude_str in args['--exclude']:
```
[the remark this answers is not in the document]
Comment (midokura-silvia) [reply treon/treon.py]: several --exclude are possible, so args['exclude'] is a list of string.

File: treon/treon.py [hunk 3]
```diff
@@ -2,13 +2,14 @@
 """
 Usage:
   treon
-  treon [PATH] [--threads=<number>] [-v]
+  treon [PATH] [--threads=<number>] [-v] [--exclude=<string>]...
```
Comment (amit1rrr) [inline treon/treon.py]: Can you please update the documentation in README as well.

File: treon/treon.py [hunk 4]
```diff
@@ -2,13 +2,14 @@
...
 
 Arguments:
   PATH                File or directory path to find notebooks to test. Searches recursively for directory paths. [default: current working directory]
 
 Options:
   --threads=<number>  Number of parallel threads. Each thread processes one notebook file at a time. [default: 10]
+  -e=<string> --exclude=<string>   Option for excluding files that start with this string. Can be provided more than once.
```
Comment (amit1rrr) [inline treon/treon.py]: Maybe reword it to something like, 

"Option for excluding files or entire directories from testing. All files whose absolute path starts with the specified string are excluded from testing. This option can be specified more than once to exclude multiple files or directories."

Your wording is also technically correct but I just wanted to hint users that this option has capability to exclude directories as well.

File: tests/test_treon.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+               'resources/runtime_error.ipynb',
+               'resources/unittest_failed.ipynb',
+               'other/resources.ipynb']
+
+    filtered = treon.filter_results(results=results, args=args)
+    expected = ['resources/doctest_failed.ipynb',
+                'resources/runtime_error.ipynb',
+                'resources/unittest_failed.ipynb',
+                'other/resources.ipynb']
+    assert filtered == expected
```
Comment (amit1rrr) [inline tests/test_treon.py]: It would be clean to write 3 separate tests, one for each assert you have - 

def exclude_option_works_with_file_paths
def exclude_option_works_with_directory_paths
def exclude_option_works_with_file_and_directory_paths_combined

that way you can move the results = [ ... ] into the setup method.

File: treon/treon.py [hunk 6]
```diff
@@ -106,6 +107,12 @@ def print_test_collection(notebooks):
     LOG.debug(message)
 
 
+def filter_results(results, args):
+    for exclude_str in args['--exclude']:
+        results = [file for file in results if not file.startswith(exclude_str)]
```
Comment (amit1rrr) [inline treon/treon.py]: Should we restrict exclude_str to be valid paths? It would be easy to implement with path module.

The benefit of that is we can handle home directory symbol("~" ) etc. also as part of exclude option. It would also look more professional to accept valid paths instead of strings.

I do not have a strong opinion on this one, so I leave it upto you to make a choice.

**Comment by amit1rrr** [comment]
Thank you for making the changes & putting in the tests. I tested with your branch & works as expected. I just have a few nitpicks after which we can merge.

File: treon/treon.py [hunk 7]
```diff
@@ -106,6 +109,13 @@ def print_test_collection(notebooks):
     LOG.debug(message)
 
 
+def filter_results(results, args):
+    for exclude_str in args['--exclude']:
+        exclude_abs = os.path.abspath(exclude_str)
```
Comment (amit1rrr) [inline treon/treon.py]: When I try os.path.abspath on python console it seems to append everything to current path.
```
>>> os.path.abspath('~/workspace')
'/Users/amitrathi/workspace/treon_forks/midokura-silvia/treon/~/workspace'
```

We might need this - 

```
>>> os.path.expanduser('~/workspace')
'/Users/amitrathi/workspace'
```

https://docs.python.org/3/library/os.path.html#os.path.expanduser

Bonus point if you could add a test for tilda (~) expansion :)

Comment (midokura-silvia) [reply treon/treon.py]: yes, I noticed this problem as well

File: README.md [hunk 8]
```diff
@@ -63,17 +63,19 @@ TEST RESULT
 ```
 Usage:
   treon
-  treon [PATH] [--threads=<number>] [-v]
+  treon [PATH] [--threads=<number>] [-v] [--exclude=<string>]...
```
Comment (amit1rrr) [inline README.md]: Let's just say,
`--exclude=<path>` instead of string here and below. To be more explicit.

File: README.md [hunk 9]
```diff
@@ -63,17 +63,19 @@ TEST RESULT
 ```
 Usage:
   treon
-  treon [PATH] [--threads=<number>] [-v]
+  treon [PATH] [--threads=<number>] [-v] [--exclude=<path>]...
```
Comment (amit1rrr) [inline README.md]: Can remove trailing `...`

Comment (midokura-silvia) [reply README.md]: The trailing ... is not a mistake. this is the syntax for letting the argument appear more than once.

Comment (amit1rrr) [reply README.md]: ah, thanks.

File: treon/treon.py [hunk 10]
```diff
@@ -106,6 +109,13 @@ def print_test_collection(notebooks):
     LOG.debug(message)
 
 
+def filter_results(results, args):
+    for exclude_str in args['--exclude']:
+        exclude_abs = os.path.abspath(os.path.expanduser(exclude_str))
+        results = [file for file in results if not os.path.abspath(file).startswith(exclude_abs)]
```
Comment (amit1rrr) [inline treon/treon.py]: I just found a problem with this, 

These are all the files I am testing normally,
```
(treon4) [amitrathi@MacBook-Pro-3 treon ((065c2f2...))]$ treon  ~/workspace/PythonDataScienceHandbook 
Executing treon version 0.1.2
Recursively scanning /Users/amitrathi/workspace/PythonDataScienceHandbook for notebooks...
Triggered test for /Users/amitrathi/workspace/PythonDataScienceHandbook/nb2/05.08-V2-Random-Forests.ipynb
Triggered test for /Users/amitrathi/workspace/PythonDataScienceHandbook/nb2/05.04-V2-Feature-Engineering.ipynb
Triggered test for /Users/amitrathi/workspace/PythonDataScienceHandbook/notebooks2/04.13v3.ipynb
Triggered test for /Users/amitrathi/workspace/PythonDataScienceHandbook/notebooks2/04.13v2.ipynb
Triggered test for /Users/amitrathi/workspace/PythonDataScienceHandbook/notebooks/03.10-Working-With-Strings.ipynb
Triggered test for /Users/amitrathi/workspace/PythonDataScienceHandbook/notebooks/05.09-Principal-Component-Analysis.ipynb
Triggered test for /Users/amitrathi/workspace/PythonDataScienceHandbook/notebooks/03.06-Concat-And-Append.ipynb
Triggered test for /Users/amitrathi/workspace/PythonDataScienceHandbook/notebooks/02.05-Computation-on-arrays-broadcasting.ipynb
Triggered test for /Users/amitrathi/workspace/PythonDataScienceHandbook/notebooks/05.05-Naive-Bayes.ipynb
Triggered test for /Users/amitrathi/workspace/PythonDataScienceHandbook/notebooks/04.14-Visualization-With-Seaborn.ipynb
```


Now when I try to ignore `......./notebooks` folder `......./notebooks2` is also ignored because of the startswith condition.
```
(treon4) [amitrathi@MacBook-Pro-3 treon ((065c2f2...))]$ treon ~/workspace/PythonDataScienceHandbook --exclude=~/workspace/PythonDataScienceHandbook/notebooks/ 
Executing treon version 0.1.2
Recursively scanning /Users/amitrathi/workspace/PythonDataScienceHandbook for notebooks...
Triggered test for /Users/amitrathi/workspace/PythonDataScienceHandbook/nb2/05.08-V2-Random-Forests.ipynb
Triggered test for /Users/amitrathi/workspace/PythonDataScienceHandbook/nb2/05.02-V2-Introducing-Scikit-Learn.ipynb
Triggered test for /Users/amitrathi/workspace/PythonDataScienceHandbook/nb2/05.04-V2-Feature-Engineering.ipynb
Triggered test for /Users/amitrathi/workspace/PythonDataScienceHandbook/nb2/03.07-Merge-and-Join.ipynb
Triggered test for /Users/amitrathi/workspace/PythonDataScienceHandbook/nb3/01.00-IPython-Beyond-Normal-Python.ipynb
```

Comment (amit1rrr) [reply treon/treon.py]: IMO there are 2 ways to solve this,

1. For each exclude path list all ipynb files under it (if path is directory) and put them in exclude set. Then exclude any file that is in the exclude set.

2. For each file check whether it is equal to exclude string OR (startswith exclude string AND exclude string has slash after the match).

Option 2 might be quick but dirty. Option 1 might be clean but inefficient. I leave it upto you.

Comment (midokura-silvia) [reply treon/treon.py]: What do you suggest?
This is a feature, I guess. Only way to avoid this is to restrict to valid directory names and valid file paths. But this can be cumbersome as well. 
I am fine with both.

Comment (amit1rrr) [reply treon/treon.py]: >Only way to avoid this is to restrict to valid directory names and valid file paths

Option 2 above would be easy to implement. We wouldn't even need to check valid paths.

Comment (amit1rrr) [reply treon/treon.py]: I edited option 2 a bit below -

For each file check whether it is equal to exclude string OR (file startswith exclude string AND filename has slash after the match).

Comment (midokura-silvia) [reply treon/treon.py]: some use the following check:
file_path.startswith(os.path.abspath(dir_path)+os.sep)

Comment (amit1rrr) [reply treon/treon.py]: Yep, something like that should do it.

**Comment by amit1rrr** [comment]
Merged! 

@midokura-silvia Thank you for your contribution!

**Comment by amit1rrr** [comment]
@midokura-silvia Released this change in `0.1.3` in case you want to start using this option right away.

https://pypi.org/project/treon/
